File: clients/haskell/src/Model/Vec2.hs
module Model.Vec2 where

import Prelude hiding (id)
import qualified Trans
import Trans (Trans)


-- | 2 dimensional vector.
data Vec2 = Vec2 {
    -- | `x` coordinate of the vector
    x :: Double,
    -- | `y` coordinate of the vector
    y :: Double }
    deriving Show

instance Trans Vec2 where
    read = do
        x <- Trans.read
        y <- Trans.read
        return Vec2 {
            x,
            y }
    
    write Vec2 {
        x,
        y } = do
            Trans.write x
            Trans.write y

File: clients/rust/src/model/vec2.rs
use super::*;

/// 2 dimensional vector.
#[derive(Clone, Debug)]
pub struct Vec2 {
    /// `x` coordinate of the vector
    pub x: f64,
    /// `y` coordinate of the vector
    pub y: f64,
}

impl trans::Trans for Vec2 {
    fn write_to(&self, writer: &mut dyn std::io::Write) -> std::io::Result<()> {
        self.x.write_to(writer)?;
        self.y.write_to(writer)?;
        Ok(())
    }
    fn read_from(reader: &mut dyn std::io::Read) -> std::io::Result<Self> {
        let x: f64 = trans::Trans::read_from(reader)?;
        let y: f64 = trans::Trans::read_from(reader)?;
        Ok(Self {
            x,
            y,
        })
    }
}